### PR TITLE
no longer emit instructions for constants

### PR DESCRIPTION
**Pull request by eellison**
Constant nodes are pooled to the beginning of the graph in the preprocessing stage. Their ivalues are written to a constants array in the Code object. At the beginning of a function call, those values are copied into the interpreter's registers.

A separate diff will be published to no longer print out constant nodes and print their ivalues inline.

### CODE REVIEW

File: torch/csrc/jit/interpreter.cpp [hunk 1]
```diff
@@ -638,10 +669,20 @@ struct CodeImpl {
...
+      bool_data(function->bool_data),
+      registers(function->register_size) {
+    loadConstantsIntoRegisters();
+  }
+
+  void loadConstantsIntoRegisters() {
+    for (auto constant: function->constants) {
+      Value * val = constant.first;
+      IValue ivalue = constant.second;
+      int reg = function->getOrAllocateRegister(val, true);
```
Comment (zdevito) [inline torch/csrc/jit/interpreter.cpp]: This is in the fast path, so it is not a good idea to do hash table lookups. Simply make the first C registers all the constants, then do a  a  std::copy of a flat vector of constants into the registers.

File: torch/csrc/jit/interpreter.cpp [hunk 2]
```diff
@@ -570,6 +575,21 @@ struct CodeImpl {
...
   }
 
+  void fillConstantsArray() {
+    auto graph_block = graph->block();
+    for (Node * n: graph_block->nodes()) {
+      //in preprocessing all constants are moved to the beginning of the graph
+      if (n->kind() != prim::Constant)
+        break;
+
+      getOrAllocateRegister(n->output());
```
Comment (zdevito) [inline torch/csrc/jit/interpreter.cpp]:
```
auto reg = getOrAllocateRegister(n->output());
JIT_ASSERT(reg == constants.size();
...
constants.emplace_back(*ivalue);
```

File: torch/csrc/jit/interpreter.cpp [hunk 3]
```diff
@@ -629,6 +656,10 @@ struct CodeImpl {
   std::vector<Instruction> instructions;
   int register_size = 0;
 
+  //we do not emit instructions for constant nodes. instead we hold their values
+  //here and copy them into the interpreter's registers when it is called
+  std::vector<std::pair<Value *, IValue>> constants;
```
Comment (zdevito) [inline torch/csrc/jit/interpreter.cpp]: `std::vector<IValue>`

**Review by zdevito — COMMENTED** [review]
One suggestion about the datastructure to use to avoid hash lookups in the fast path. Otherwise this looks good.

File: torch/csrc/jit/graph_executor.cpp [hunk 4]
```diff
@@ -445,11 +445,11 @@ struct GraphExecutorImpl {
   void runOptimization(std::shared_ptr<Graph>& graph, const ArgumentSpec& spec) {
     EliminateDeadCode(graph);
     EliminateCommonSubexpression(graph);
-    ConstantPooling(graph);
     UnrollLoops(graph);
     PeepholeOptimize(graph);
     CheckInplace(graph);
     BatchMM(graph);
+    ConstantPooling(graph);
```
Comment (apaszke) [inline torch/csrc/jit/graph_executor.cpp]: You raised a valid point when you said that constants will count towards the size of loop in loop unrolling, and so we should use that before that happens. It should be ok to call this pass twice.

File: torch/csrc/jit/interpreter.cpp (hunk 1 above)
Comment (apaszke) [inline torch/csrc/jit/interpreter.cpp]: This shouldn't be ever called at runtime. Just put their indices into the flattened int vector that we already have allocated, and use those (just like we do for all other register accesses).

File: torch/csrc/jit/graph_executor.cpp (hunk 4 above)
Comment (apaszke) [reply torch/csrc/jit/graph_executor.cpp]: Oh I see you changed loop unrolling. Never mind my comment.

**Pull request closed without merging**